rust/kernel/chrdev.rs

```diff
@@ -20,10 +20,66 @@ use crate::error::{Error, KernelResult};
 use crate::file_operations;
 use crate::types::CStr;
 
+struct SafeCdev(*mut bindings::cdev);
+
+impl SafeCdev {
+    fn alloc() -> KernelResult<Self> {
+        // SAFETY: call an unsafe function
+        let cdev = unsafe { bindings::cdev_alloc() };
+        if cdev.is_null() {
+            return Err(Error::ENOMEM);
+        }
+        Ok(Self(cdev))
+    }
+
+    fn init(&mut self, fops: *const bindings::file_operations) {
+        // SAFETY: call an unsafe function
+        unsafe {
+            bindings::cdev_init(self.0, fops);
```

Comment:
I don't think we're supposed to call `cdev_init` on what's returned by `cdev_alloc `. Perhaps we should have the constructor above (`alloc`) also take a `file_operations` as argument and we just initialise `cdev.ops`?

Reply:
You are absolutely right, that pattern is incorrect! Yes, let's fold the `cdev.ops` assignment into `alloc`.